File: dyros_robot_controller/src/manipulator/robot_controller.cpp
#include "dyros_robot_controller/manipulator/robot_controller.h"

namespace drc
{
    namespace Manipulator
    {
        RobotController::RobotController(const double& dt,
                                         std::shared_ptr<Manipulator::RobotData> robot_data)
        : robot_data_(robot_data)
        {
            dof_ = robot_data_->getDof();
            Kp_task_ = VectorXd::Constant(6, 100);
            Kv_task_ = VectorXd::Constant(6, 20);
            Kp_joint_ = VectorXd::Constant(dof_, 400);
            Kv_joint_ = VectorXd::Constant(dof_, 40);
    
            QP_mani_IK_ = std::make_unique<Manipulator::QPIK>(robot_data_);
            QP_mani_ID_ = std::make_unique<Manipulator::QPID>(robot_data_);
        }
    
        void RobotController::setJointGain(const VectorXd& Kp, const VectorXd& Kv)
        {
            if (Kp.size() != dof_ || Kv.size() != dof_)
            {
                throw std::runtime_error("Kp and Kv must be of size dof_.");
            }
            Kp_joint_ = Kp;
            Kv_joint_ = Kv;
        }

        void RobotController::setJointKpGain(const VectorXd& Kp)
        {
            if (Kp.size() != dof_ )
            {
                throw std::runtime_error("Kp must be of size dof_.");
            }
            Kp_joint_ = Kp;
        }

        void RobotController::setJointKvGain(const VectorXd& Kv)
        {
            if (Kv.size() != dof_ )
            {
                throw std::runtime_error("Kv must be of size dof_.");
            }
            Kv_joint_ = Kv;
        }
    
        void RobotController::setTaskGain(const VectorXd& Kp, const VectorXd& Kv)
        {
            if (Kp.size() != 6 || Kv.size() != 6)
            {
                throw std::runtime_error("Kp and Kv must be of size 6.");
            }
            Kp_task_ = Kp;
            Kv_task_ = Kv;
        }

        void RobotController::setTaskKpGain(const VectorXd& Kp)
        {
            if (Kp.size() != 6)
            {
                throw std::runtime_error("Kp must be of size 6.");
            }
            Kp_task_ = Kp;
        }

        void RobotController::setTaskKvGain(const VectorXd& Kv)
        {
            if (Kv.size() != 6)
            {
                throw std::runtime_error("Kv must be of size 6.");
            }
            Kv_task_ = Kv;
        }
    
        VectorXd RobotController::moveJointPositionCubic(const VectorXd& q_target,
                                                         const VectorXd& qdot_target,
                                                         const VectorXd& q_init,
                                                         const VectorXd& qdot_init,
                                                         const double& current_time,
                                                         const double& init_time,
                                                         const double& duration)
        {
            VectorXd q_desired =  DyrosMath::cubicVector(current_time,
                                                         init_time,
                                                         init_time + duration,
                                                         q_init,
                                                         q_target,
                                                         qdot_init,
                                                         qdot_target);
    
            return q_desired;
        }

        VectorXd RobotController::moveJointVelocityCubic(const VectorXd& q_target,
                                                         const VectorXd& qdot_target,
                                                         const VectorXd& q_init,
                                                         const VectorXd& qdot_init,
                                                         const double& current_time,
                                                         const double& init_time,
                                                         const double& duration)
        {
            VectorXd qdot_desired =  DyrosMath::cubicDotVector(current_time,
                                                               init_time,
                                                               init_time + duration,
                                                               q_init,
                                                               q_target,
                                                               qdot_init,
                                                               qdot_target);
    
            return qdot_desired;
        }

        VectorXd RobotController::moveJointTorqueStep(const VectorXd& qddot_target)
        {
            return robot_data_->getMassMatrix() * qddot_target + robot_data_->getGravity();;
        }

        VectorXd RobotController::moveJointTorqueStep(const VectorXd& q_target,
                                                      const VectorXd& qdot_target)
        {
            VectorXd qddot_desired = Kp_joint_.asDiagonal() * (q_target - robot_data_->getJointPosition()) + Kv_joint_.asDiagonal() * (qdot_target - robot_data_->getJointVelocity());
            return moveJointTorqueStep(qddot_desired);
        }
    
        VectorXd RobotController::moveJointTorqueCubic(const VectorXd& q_target,
                                                       const VectorXd& qdot_target,
                                                       const VectorXd& q_init,
                                                       const VectorXd& qdot_init,
                                                       const double& current_time,
                                                       const double& init_time,
                                                       const double& duration)
        {
            VectorXd q_desired =  DyrosMath::cubicVector(current_time,
                                                         init_time,
                                                         init_time + duration,
                                                         q_init,
                                                         q_target,
                                                         qdot_init,
                                                         qdot_target);
    
            VectorXd qdot_desired =  DyrosMath::cubicDotVector(current_time,
                                                               init_time,
                                                               init_time + duration,
                                                               q_init,
                                                               q_target,
                                                               qdot_init,
                                                               qdot_target);
    
            return moveJointTorqueStep(q_desired, qdot_desired);
        }
    
        
    
        VectorXd RobotController::CLIKStep(const Affine3d& x_target,
                                           const VectorXd& xdot_target,
                                           const VectorXd& null_qdot,
                                           const std::string& link_name)
        {
            VectorXd x_error, xdot_error;
            DyrosMath::getTaskSpaceError(x_target, xdot_target, robot_data_->getPose(link_name), robot_data_->getVelocity(link_name), x_error, xdot_error);
    
            MatrixXd J = robot_data_->getJacobian(link_name);
            MatrixXd J_pinv = DyrosMath::PinvCOD(J);
    
            MatrixXd null_proj = MatrixXd::Identity(dof_, dof_) - J_pinv * J;
    
            VectorXd qdot_desired = J_pinv * (Kp_task_.asDiagonal() * x_error + xdot_target) + null_proj * null_qdot;
            return qdot_desired;
        }

        VectorXd RobotController::CLIKStep(const Affine3d& x_target,
                                           const VectorXd& xdot_target,
                                           const std::string& link_name)
        {
            return CLIKStep(x_target, xdot_target, VectorXd::Zero(dof_), link_name);
        }
    
        VectorXd RobotController::CLIKCubic(const Affine3d& x_target,
                                            const VectorXd& xdot_target,
                                            const Affine3d& x_init,
                                            const VectorXd& xdot_init,
                                            const double& current_time,
                                            const double& init_time,
                                            const double& duration,
                                            const VectorXd& null_qdot,
                                            const std::string& link_name)
        {
            Affine3d x_desired;
            VectorXd xdot_desired;
            DyrosMath::getTaskSpaceCubic(x_target, xdot_target, x_init, xdot_init, current_time, init_time, duration, x_desired, xdot_desired);
            return CLIKStep(x_desired, xdot_desired, null_qdot, link_name);
        }

        VectorXd RobotController::CLIKCubic(const Affine3d& x_target,
                                            const VectorXd& xdot_target,
                                            const Affine3d& x_init,
                                            const VectorXd& xdot_init,
                                            const double& current_time,
                                            const double& init_time,
                                            const double& duration,
                                            const std::string& link_name)
        {
            return CLIKCubic(x_target, xdot_target, x_init, xdot_init, current_time, init_time, duration, VectorXd::Zero(dof_), link_name);
        }
    
        VectorXd RobotController::OSF(const VectorXd& xddot_target, 
                                      const VectorXd& null_torque,
                                      const std::string& link_name)
        {
            MatrixXd J = robot_data_->getJacobian(link_name);
            MatrixXd J_T = J.transpose();
            MatrixXd M_inv = robot_data_->getMassMatrixInv();
    
            MatrixXd M_task = DyrosMath::PinvCOD(J * M_inv * J_T); 
            MatrixXd J_T_pinv = M_task * J * M_inv;
            MatrixXd null_proj = MatrixXd::Identity(dof_, dof_) - (J_T * J_T_pinv);
            
            VectorXd force_desired = M_task * xddot_target;
            
            return J_T * force_desired + null_proj * null_torque + robot_data_->getGravity();
        }
    
        VectorXd RobotController::OSF(const VectorXd& xddot_target,
                                      const std::string& link_name)
        {
            return OSF(xddot_target, VectorXd::Zero(dof_), link_name);
        }
    
        VectorXd RobotController::OSFStep(const Affine3d& x_target, 
                                          const VectorXd& xdot_target,
                                          const VectorXd& null_torque,
                                          const std::string& link_name)
        {
            VectorXd x_error, xdot_error;
            DyrosMath::getTaskSpaceError(x_target, xdot_target, robot_data_->getPose(link_name), robot_data_->getVelocity(link_name), x_error, xdot_error);
            VectorXd xddot_target = Kp_task_.asDiagonal() * x_error + Kv_task_.asDiagonal() * xdot_error;
            return OSF(xddot_target, null_torque, link_name);
        }
    
        VectorXd RobotController::OSFStep(const Affine3d& x_target, 
                                          const VectorXd& xdot_target,
                                          const std::string& link_name)
        {
            return OSFStep(x_target, xdot_target, VectorXd::Zero(dof_), link_name);
        }
    
        VectorXd RobotController::OSFCubic(const Affine3d& x_target,
                                           const VectorXd& xdot_target,
                                           const Affine3d& x_init,
                                           const VectorXd& xdot_init,
                                           const double& current_time,
                                           const double& init_time,
                                           const double& duration,
                                           const VectorXd& null_torque,
                                           const std::string& link_name)
        {
            Affine3d x_desired;
            VectorXd xdot_desired;
            DyrosMath::getTaskSpaceCubic(x_target, xdot_target, x_init, xdot_init, current_time, init_time, duration, x_desired, xdot_desired);
            return OSFStep(x_desired, xdot_desired, null_torque, link_name);
        }
    
        VectorXd RobotController::OSFCubic(const Affine3d& x_target,
                                           const VectorXd& xdot_target,
                                           const Affine3d& x_init,
                                           const VectorXd& xdot_init,
                                           const double& current_time,
                                           const double& init_time,
                                           const double& duration,
                                           const std::string& link_name)
        {
            return OSFCubic(x_target, xdot_target, x_init, xdot_init, current_time, init_time, duration, VectorXd::Zero(dof_), link_name);
        }
    
        VectorXd RobotController::QPIK(const VectorXd& xdot_target,
                                       const std::string& link_name)
        {  
            QP_mani_IK_->setDesiredTaskVel(xdot_target, link_name);
            VectorXd opt_qdot = VectorXd::Zero(dof_);
            QP::TimeDuration tmp;
            if(!QP_mani_IK_->getOptJointVel(opt_qdot, tmp))
            {
                std::cerr << "QP IK failed to compute optimal joint velocity." << std::endl;
                opt_qdot.setZero(dof_);
            }
        
            return opt_qdot;
        }

        VectorXd RobotController::QPIKStep(const Affine3d& x_target, 
                                           const VectorXd& xdot_target,
                                           const std::string& link_name)
        {
            VectorXd x_error, xdot_error;
            DyrosMath::getTaskSpaceError(x_target, xdot_target, robot_data_->getPose(link_name), robot_data_->getVelocity(link_name), x_error, xdot_error);
            
            return QPIK(Kp_task_.asDiagonal() * x_error + Kv_task_.asDiagonal() * xdot_error, link_name);
        }
    
    
        VectorXd RobotController::QPIKCubic(const Affine3d& x_target,
                                            const VectorXd& xdot_target,
                                            const Affine3d& x_init,
                                            const VectorXd& xdot_init,
                                            const double& current_time,
                                            const double& init_time,
                                            const double& duration,
                                            const std::string& link_name)
        {
            Affine3d x_desired;
            VectorXd xdot_desired;
            DyrosMath::getTaskSpaceCubic(x_target, xdot_target, x_init, xdot_init, current_time, init_time, duration, x_desired, xdot_desired);
    
            return QPIKStep(x_desired, xdot_desired, link_name);
        }

        VectorXd RobotController::QPID(const VectorXd& xddot_target,
                                       const std::string& link_name)
        {
            QP_mani_ID_->setDesiredTaskAcc(xddot_target, link_name);
            VectorXd opt_qddot = VectorXd::Zero(dof_);
            VectorXd opt_torque = VectorXd::Zero(dof_);
            QP::TimeDuration tmp;
            if(!QP_mani_ID_->getOptJoint(opt_qddot, opt_torque, tmp))
            {
                std::cerr << "QP ID failed to compute optimal joint torque." << std::endl;
                opt_torque = robot_data_->getGravity();
            }

            return opt_torque;
        }

        VectorXd RobotController::QPIDStep(const Affine3d& x_target, 
                                           const VectorXd& xdot_target,
                                           const std::string& link_name)
        {
            VectorXd x_error, xdot_error;
            DyrosMath::getTaskSpaceError(x_target, xdot_target, robot_data_->getPose(link_name), robot_data_->getVelocity(link_name), x_error, xdot_error);
            
            return QPID(Kp_task_.asDiagonal() * x_error + Kv_task_.asDiagonal() * xdot_error, link_name);
        }
    
    
        VectorXd RobotController::QPIDCubic(const Affine3d& x_target,
                                            const VectorXd& xdot_target,
                                            const Affine3d& x_init,
                                            const VectorXd& xdot_init,
                                            const double& current_time,
                                            const double& init_time,
                                            const double& duration,
                                            const std::string& link_name)
        {
            Affine3d x_desired;
            VectorXd xdot_desired;
            DyrosMath::getTaskSpaceCubic(x_target, xdot_target, x_init, xdot_init, current_time, init_time, duration, x_desired, xdot_desired);
    
            return QPIDStep(x_desired, xdot_desired, link_name);
        }
    } // namespace Manipulator
} // namespace drc

File: dyros_robot_controller/drc/manipulator/robot_controller.py
import numpy as np
import dyros_robot_controller_cpp_wrapper as drc
from .robot_data import RobotData

class RobotController(drc.ManipulatorRobotController):
    """
    A Python wrapper for the C++ RobotController::Manipulator::ManipulatorBase class.
    
    This class consists of functions that compute manipulator control inputs and helpers that generate smooth trajectories.
    Joint space functions compute control inputs to track desired joint positions, velocities, or accelerations.
    Task space functions compute control inputs to track desired position, velocity, or acceleration of a link.
    """
    def __init__(self, dt: float, robot_data: RobotData):
        """
        Constructor.

        Parameters:
            dt         : (float) Control loop time step in seconds.
            robot_data : (DataManipulatorBase) An instance of the Python ManipulatorBase wrapper which contains the robot's kinematic and dynamic parameters.

        Raises:
            TypeError: If the robot_data is not an instance of the Python ManipulatorBase wrapper.
        """
        if not isinstance(robot_data, RobotData):
            raise TypeError("robot_data must be an instance of the Python ManipulatorBase wrapper")

        self._dt = float(dt)
        self._robot_data = robot_data
        super().__init__(self._dt, self._robot_data)

    def set_joint_gain(self, kp: np.ndarray, kv: np.ndarray):
        """
        Set joint space PD gains for the manipulator.

        Parameters:
            kp : (np.ndarray) Proportional gains.
            kv : (np.ndarray) Derivative gains.
        """
        super().setJointGain(kp, kv)
        
    def set_joint_kp_gain(self, kp: np.ndarray):
        """
        Set joint space P gains for the manipulator.

        Parameters:
            kp : (np.ndarray) Proportional gains.
        """
        super().setJointKpGain(kp)
        
    def set_joint_kv_gain(self, kv: np.ndarray):
        """
        Set joint space D gains for the manipulator.

        Parameters:
            kv : (np.ndarray) Derivative gains.
        """
        super().setJointKvGain(kv)

    def set_task_gain(self, kp: np.ndarray, kv: np.ndarray):
        """
        Set task space PD gains for the manipulator.

        Parameters:
            kp : (np.ndarray) Proportional gains.
            kv : (np.ndarray) Derivative gains.
        """
        super().setTaskGain(kp, kv)
        
    def set_task_kp_gain(self, kp: np.ndarray):
        """
        Set task space P gains for the manipulator.

        Parameters:
            kp : (np.ndarray) Proportional gains.
        """
        super().setTaskKpGain(kp)
        
    def set_task_kv_gain(self, kv: np.ndarray):
        """
        Set task space D gains for the manipulator.

        Parameters:
            kv : (np.ndarray) Derivative gains.
        """
        super().setTaskKpGain(kv)

    # ================================ Joint space Functions ================================
    def move_joint_position_cubic(self,
                                  q_target: np.ndarray,
                                  qdot_target: np.ndarray,
                                  q_init: np.ndarray,
                                  qdot_init: np.ndarray,
                                  current_time: float,
                                  init_time: float,
                                  duration: float,
                                  ) -> np.ndarray:
        """
        Perform cubic interpolation between the initial and desired joint configurations over the given duration.

        Parameters:
            q_target     : (np.ndarray) Desired manipulator joint positions at the end of the segment.
            qdot_target  : (np.ndarray) Desired manipulator joint velocities at the end of the segment.
            q_init       : (np.ndarray) Initial manipulator joint positions at the start of the segment.
            qdot_init    : (np.ndarray) Initial manipulator joint velocities at the start of the segment.
            current_time : (float) Current time.
            init_time    : (float) Start time of the segment.
            duration     : (float) Time duration.

        Returns:
            (np.ndarray) Desired joint positions.
        """
        return super().moveJointPositionCubic(q_target,
                                              qdot_target,
                                              q_init,
                                              qdot_init,
                                              current_time,
                                              init_time,
                                              duration, 
                                              )
        
    def move_joint_velocity_cubic(self,
                                  q_target: np.ndarray,
                                  qdot_target: np.ndarray,
                                  q_init: np.ndarray,
                                  qdot_init: np.ndarray,
                                  current_time: float,
                                  init_time: float,
                                  duration: float,
                                  ) -> np.ndarray:
        """
        Perform cubic interpolation between the initial and desired joint configurations over the given duration.

        Parameters:
            q_target     : (np.ndarray) Desired manipulator joint positions at the end of the segment.
            qdot_target  : (np.ndarray) Desired manipulator joint velocities at the end of the segment.
            q_init       : (np.ndarray) Initial manipulator joint positions at the start of the segment.
            qdot_init    : (np.ndarray) Initial manipulator joint velocities at the start of the segment.
            current_time : (float) Current time.
            init_time    : (float) Start time of the segment.
            duration     : (float) Time duration.

        Returns:
            (np.ndarray) Desired joint velocities.
        """
        return super().moveJointVelocityCubic(q_target,
                                              qdot_target,
                                              q_init,
                                              qdot_init,
                                              current_time,
                                              init_time,
                                              duration, 
                                              )

    # def move_joint_torque_step(self, q_target: np.ndarray, qdot_target: np.ndarray) -> np.ndarray:
    #     return super().moveJointTorqueStep(q_target, qdot_target)
    
    # def move_joint_torque_step(self, qddot_target: np.ndarray) -> np.ndarray:
    #     return super().moveJointTorqueStep(qddot_target)
    
    def move_joint_torque_step(self,
                               q_target:     np.ndarray | None = None,
                               qdot_target:  np.ndarray | None = None,
                               qddot_target: np.ndarray | None = None,
                               ) -> np.ndarray:
        """
        Computes joint torques to achieve desired joint configurations using equations of motion and PD control law.

        Parameters:
            q_target     : (np.ndarray) [Required if qddot_target is None]
                            Desired manipulator joint positions.
            qdot_target  : (np.ndarray) [Required if qddot_target is None]
                            Desired manipulator joint velocities.
            qddot_target : (np.ndarray) [Required if q_target and qdot_target are None]
                            Desired manipulator joint accelerations.

        Returns:
            (np.ndarray) Desired joint torques.
        """
        if qddot_target is not None:
            return super().moveJointTorqueStep(qddot_target)

        if q_target is not None and qdot_target is not None:
            return super().moveJointTorqueStep(q_target, qdot_target)

    def move_joint_torque_cubic(self,
                                q_target: np.ndarray,
                                qdot_target: np.ndarray,
                                q_init: np.ndarray,
                                qdot_init: np.ndarray,
                                current_time: float,
                                init_time: float,
                                duration: float,
                                ) -> np.ndarray:
        """
        Perform cubic interpolation between the initial and desired joint configurations over the given duration, then compute joint torques to follow the resulting trajectory.

        Parameters:
            q_target     : (np.ndarray) Desired manipulator joint positions at the end of the segment.
            qdot_target  : (np.ndarray) Desired manipulator joint velocities at the end of the segment.
            q_init       : (np.ndarray) Initial manipulator joint positions at the start of the segment.
            qdot_init    : (np.ndarray) Initial manipulator joint velocities at the start of the segment.
            current_time : (float) Current time.
            init_time    : (float) Start time of the segment.
            duration     : (float) Time duration.

        Returns:
            (np.ndarray) Desired joint torques.
        """
        return super().moveJointTorqueCubic(q_target,
                                            qdot_target,
                                            q_init,
                                            qdot_init,
                                            current_time,
                                            init_time,
                                            duration,
                                            )

    # ================================ Task space Functions ================================
    def CLIK_step(self,
                  x_target: np.ndarray,
                  xdot_target: np.ndarray,
                  link_name: str,
                  null_qdot: np.ndarray | None = None,
                  ) -> np.ndarray:
        """
        Computes joint velocity to achieve desired position & velocity of a link using closed-loop inverse kinematics, projecting null_qdot into null space to exploit redundancy if provided.

        Parameters:
            x_target    : (np.ndarray) Desired position of a link.
            xdot_target : (np.ndarray) Desired velocity of a link.
            link_name   : (str) Name of the link.
            null_qdot   : (np.ndarray) [Optional] Desired joint velocity to be projected on null space.

        Returns:
            (np.ndarray) Desired joint velocities.
        """
        if null_qdot is None:
            return super().CLIKStep(x_target,
                                    xdot_target,
                                    link_name,
                                    )
        else:
            return super().CLIKStep(x_target,
                                    xdot_target,
                                    null_qdot,
                                    link_name,
                                    )

    def CLIK_cubic(self,
                   x_target: np.ndarray,
                   xdot_target: np.ndarray,
                   x_init: np.ndarray,
                   xdot_init: np.ndarray,
                   current_time: float,
                   init_time: float,
                   duration: float,
                   link_name: str,
                   null_qdot: np.ndarray | None = None,
                   ) -> np.ndarray:
        """
        Perform cubic interpolation between the initial and desired link pose and velocity over the given duration, then compute joint velocities to follow the resulting trajectory.

        Parameters:
            x_target     : (np.ndarray) Desired position of a link at the end of the segment.
            xdot_target  : (np.ndarray) Desired velocity of a link at the end of the segment.
            x_init       : (np.ndarray) Initial position of a link at the start of the segment.
            xdot_init    : (np.ndarray) Initial velocity of a link at the start of the segment.
            current_time : (float) Current time.
            init_time    : (float) Start time of the segment.
            duration     : (float) Time duration.
            link_name    : (str) Name of the link.
            null_qdot    : (np.ndarray) [Optional] Desired joint velocity to be projected on null space.

        Returns:
            (np.ndarray) Desired joint velocities.
        """
        if null_qdot is None:
            return super().CLIKCubic(x_target, 
                                     xdot_target,
                                     x_init,   
                                     xdot_init,
                                     current_time, 
                                     init_time, 
                                     duration,
                                     link_name
                                     )
        else:
            return super().CLIKCubic(x_target, 
                                     xdot_target,
                                     x_init,   
                                     xdot_init,
                                     current_time, 
                                     init_time, 
                                     duration,
                                     null_qdot,
                                     link_name
                                     )


    def OSF(self,
            xddot_target: np.ndarray,
            link_name: str,
            null_torque: np.ndarray | None = None,
            ) -> np.ndarray:
        """
        Computes joint torque to achieve desired acceleration of a link using operational space control, projecting null_torque into null space to exploit redundancy if provided.

        Parameters:
            xddot_target : (np.ndarray) Desired acceleration of a link.
            link_name    : (str) Name of the link.
            null_torque  : (np.ndarray) [Optional] Desired joint torque to be projected on null space.

        Returns:
            (np.ndarray) Desired joint torques.
        """
        if null_torque is None:
            return super().OSF(xddot_target, link_name)
        else:
            return super().OSF(xddot_target, null_torque, link_name)

    def OSF_step(self,
                 x_target: np.ndarray,
                 xdot_target: np.ndarray,
                 link_name: str,
                 null_torque: np.ndarray | None = None,
                 ) -> np.ndarray:
        """
        Computes joint torque to achieve desired position & velocity of a link using operational space control, projecting null_torque into null space to exploit redundancy if provided.

        Parameters:
            x_target    : (np.ndarray) Desired position of a link.
            xdot_target : (np.ndarray) Desired velocity of a link.
            link_name   : (str) Name of the link.
            null_torque : (np.ndarray) [Optional] Desired joint torque to be projected on null space.

        Returns:
            (np.ndarray) Desired joint torques.
        """
        if null_torque is None:
            return super().OSFStep(x_target, xdot_target, link_name)
        else: 
            return super().OSFStep(x_target, xdot_target, null_torque, link_name)

    def OSF_cubic(self,
                  x_target: np.ndarray,
                  xdot_target: np.ndarray,
                  x_init: np.ndarray,
                  xdot_init: np.ndarray,
                  current_time: float,
                  init_time: float,
                  duration: float,
                  link_name: str,
                  null_torque: np.ndarray | None = None,
                  ) -> np.ndarray:
        """
        Perform cubic interpolation between the initial and desired link pose and velocity over the given duration, then compute joint torques to follow the resulting trajectory.

        Parameters:
            x_target     : (np.ndarray) Desired position of a link at the end of the segment.
            xdot_target  : (np.ndarray) Desired velocity of a link at the end of the segment.
            x_init       : (np.ndarray) Initial position of a link at the start of the segment.
            xdot_init    : (np.ndarray) Initial velocity of a link at the start of the segment.
            current_time : (float) Current time.
            init_time    : (float) Start time of the segment.
            duration     : (float) Time duration.
            link_name    : (str) Name of the link.
            null_torque  : (np.ndarray) [Optional] Desired joint torque to be projected on null space.

        Returns:
            (np.ndarray) Desired joint torques.
        """
        if null_torque is None:
            return super().OSFCubic(x_target,
                                    xdot_target,
                                    x_init,
                                    xdot_init,
                                    current_time,
                                    init_time,
                                    duration,
                                    link_name,
                                    )
        else:
            return super().OSFCubic(x_target,
                                    xdot_target,
                                    x_init,
                                    xdot_init,
                                    current_time,
                                    init_time,
                                    duration,
                                    null_torque,
                                    link_name,
                                    )
    def QPIK(self, xdot_target: np.ndarray, link_name: str) -> np.ndarray:
        """
        Computes joint velocities to achieve desired velocity of a link by solving inverse kinematics QP.

        Parameters:
            xdot_target : (np.ndarray) Desired velocity of a link.
            link_name   : (str) Name of the link.

        Returns:
            (np.ndarray) Desired joint velocities.
        """
        return super().QPIK(xdot_target, link_name)

    def QPIK_step(self, x_target: np.ndarray, xdot_target: np.ndarray, link_name: str) -> np.ndarray:
        """
        Computes joint velocities to achieve desired position & velocity of a link by solving inverse kinematics QP.

        Parameters:
            x_target    : (np.ndarray) Desired position of a link.
            xdot_target : (np.ndarray) Desired velocity of a link.
            link_name   : (str) Name of the link.

        Returns:
            (np.ndarray) Desired joint velocities.
        """
        return super().QPIKStep(x_target, xdot_target, link_name)

    def QPIK_cubic(self,
                   x_target: np.ndarray,
                   xdot_target: np.ndarray,
                   x_init: np.ndarray,
                   xdot_init: np.ndarray,
                   current_time: float,
                   init_time: float,
                   duration: float,
                   link_name: str,
                   ) -> np.ndarray:
        """
        Perform cubic interpolation between the initial and desired link pose & velocity over the given duration, then compute joint velocities using QP to follow the resulting trajectory.

        Parameters:
            x_target     : (np.ndarray) Desired position of a link at the end of the segment.
            xdot_target  : (np.ndarray) Desired velocity of a link at the end of the segment.
            x_init       : (np.ndarray) Initial position of a link at the start of the segment.
            xdot_init    : (np.ndarray) Initial velocity of a link at the start of the segment.
            current_time : (float) Current time.
            init_time    : (float) Start time of the segment.
            duration     : (float) Time duration.
            link_name    : (str) Name of the link.

        Returns:
            (np.ndarray) Desired joint velocities.
        """
        return super().QPIKCubic(x_target,
                                 xdot_target,
                                 x_init,
                                 xdot_init,
                                 current_time,
                                 init_time,
                                 duration,
                                 link_name,
                                 )

    def QPID(self, xddot_target: np.ndarray, link_name: str) -> np.ndarray:
        """
        Computes joint torques to achieve desired acceleration of a link by solving inverse dynamics QP.

        Parameters:
            xddot_target : (np.ndarray) Desired acceleration of a link.
            link_name    : (str) Name of the link.

        Returns:
            (np.ndarray) Desired joint torques.
        """
        return super().QPID(xddot_target, link_name)

    def QPID_step( self, x_target: np.ndarray, xdot_target: np.ndarray, link_name: str) -> np.ndarray:
        """
        Computes joint torques to achieve desired position & velocity of a link by solving inverse dynamics QP.

        Parameters:
            x_target    : (np.ndarray) Desired position of a link.
            xdot_target : (np.ndarray) Desired velocity of a link.
            link_name   : (str) Name of the link.

        Returns:
            (np.ndarray) Desired joint torques.
        """
        return super().QPIDStep(x_target, xdot_target, link_name)

    def QPID_cubic(self,
                   x_target: np.ndarray,
                   xdot_target: np.ndarray,
                   x_init: np.ndarray,
                   xdot_init: np.ndarray,
                   current_time: float,
                   init_time: float,
                   duration: float,
                   link_name: str,
                   ) -> np.ndarray:
        """
        Perform cubic interpolation between the initial and desired link pose & velocity over the given duration, then compute joint torques using QP to follow the resulting trajectory.

        Parameters:
            x_target     : (np.ndarray) Desired position of a link at the end of the segment.
            xdot_target  : (np.ndarray) Desired velocity of a link at the end of the segment.
            x_init       : (np.ndarray) Initial position of a link at the start of the segment.
            xdot_init    : (np.ndarray) Initial velocity of a link at the start of the segment.
            current_time : (float) Current time.
            init_time    : (float) Start time of the segment.
            duration     : (float) Time duration.
            link_name    : (str) Name of the link.

        Returns:
            (np.ndarray) Desired joint torques.
        """
        return super().QPIDCubic(x_target,
                                 xdot_target,
                                 x_init,
                                 xdot_init,
                                 current_time,
                                 init_time,
                                 duration,
                                 link_name,
                                 )